channeldb/payments.go

```diff
@@ -67,267 +142,310 @@ func (ps *PaymentStatus) FromBytes(status []byte) error {
+	// NOTE: Can be nil if no attempt is yet made.
+	Attempt *PaymentAttemptInfo
 
-		// We use BigEndian for keys as it orders keys in
-		// ascending order. This allows bucket scans to order payments
-		// in the order in which they were created.
-		paymentIDBytes := make([]byte, 8)
-		binary.BigEndian.PutUint64(paymentIDBytes, paymentID)
+	// PaymentPreimage is the preimage of a successful payment. This serves
+	// as a proof of payment. It will only be non-nil for settled payments.
+	//
+	// NOTE: Can be nil if payment is not settled.
+	PaymentPreimage *lntypes.Preimage
 
-		return payments.Put(paymentIDBytes, paymentBytes)
-	})
+	// Failure is a failure reason code indicating the reason the payment
+	// failed. It is only non-nil for failed payments.
+	//
+	// NOTE: Can be nil if payment is not failed.
+	Failure *FailureReason
 }
 
-// FetchAllPayments returns all outgoing payments in DB.
-func (db *DB) FetchAllPayments() ([]*OutgoingPayment, error) {
-	var payments []*OutgoingPayment
+// FetchPayments returns all sent payments found in the DB.
+func (db *DB) FetchPayments() ([]*Payment, error) {
+	var payments []*Payment
 
 	err := db.View(func(tx *bbolt.Tx) error {
-		bucket := tx.Bucket(paymentBucket)
-		if bucket == nil {
-			return ErrNoPaymentsCreated
+		paymentsBucket := tx.Bucket(paymentsRootBucket)
+		if paymentsBucket == nil {
+			return nil
 		}
 
-		return bucket.ForEach(func(k, v []byte) error {
-			// If the value is nil, then we ignore it as it may be
-			// a sub-bucket.
-			if v == nil {
-				return nil
+		return paymentsBucket.ForEach(func(k, v []byte) error {
+			bucket := paymentsBucket.Bucket(k)
+			if bucket == nil {
+				// We only expect sub-buckets to be found in
+				// this top-level bucket.
+				return fmt.Errorf("non bucket element in " +
+					"payments bucket")
 			}
 
-			r := bytes.NewReader(v)
-			payment, err := deserializeOutgoingPayment(r)
+			p, err := fetchPayment(bucket)
 			if err != nil {
 				return err
 			}
 
-			payments = append(payments, payment)
-			return nil
+			payments = append(payments, p)
+
+			// For older versions of lnd, duplicate payments to a
+			// payment has was possible. These will be found in a
+			// sub-bucket indexed by their sequence number if
+			// available.
+			dup := bucket.Bucket(paymentDuplicateBucket)
+			if dup == nil {
+				return nil
+			}
+
+			return dup.ForEach(func(k, v []byte) error {
+				subBucket := dup.Bucket(k)
+				if subBucket == nil {
+					// We one bucket for each duplicate to
+					// be found.
+					return fmt.Errorf("non bucket element" +
+						"in duplicate bucket")
+				}
+
+				p, err := fetchPayment(subBucket)
+				if err != nil {
+					return err
+				}
+
+				payments = append(payments, p)
+				return nil
+			})
 		})
 	})
 	if err != nil {
 		return nil, err
 	}
 
+	// Before returning, sort the payments by their sequence number.
+	sort.Slice(payments, func(i, j int) bool {
+		return payments[i].sequenceNum < payments[j].sequenceNum
+	})
```

Comment:
Is this a presentation only thing? If so, consider moving to rpcserver.go

Reply:
Yes, kept it here to keep the API similar to before. Can consider moving it if we want to expose the sequence number outside this package.

Reply:
This is not external api, so moving to `rpcserver` would change nothing really? (non blocking)

Reply:
Only change would be to making `sequenceNum` exported.

Reply:
Oh, ok, see what you mean now. Then leave as is indeed.